### PR TITLE
fix: list_indices never updated after first call

### PR DESCRIPTION
**Pull request by BubbleCal**
lance in rust would cache the index metadata so no need to cache it in python.

- also added an option to control which indices to optimize
- also added tests for optimize_indices

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -2790,6 +2788,15 @@ def optimize_indices(self, **kwargs):
         the new data to existing partitions.  This means an update is much quicker
         than retraining the entire index but may have less accuracy (especially
         if the new data exhibits new patterns, concepts, or trends)
+
+        Parameters
+        ----------
+        num_indices_to_merge: int, default 1
+            The number of indices to merge.
+            If set to 0, new delta index will be created.
+        indexe_names: List[str], default None
```
Comment (westonpace) [inline python/python/lance/dataset.py:2797]:
```suggestion
        index_names: List[str], default None
```

File: rust/lance/src/index.rs [hunk 2]
```diff
@@ -1019,9 +1028,25 @@ mod tests {
         assert_eq!(stats["num_unindexed_fragments"], 1);
         assert_eq!(stats["num_indices"], 1);
 
+        dataset
+            .optimize_indices(&OptimizeOptions {
+                num_indices_to_merge: 0,   // Just create index for delta
+                index_names: Some(vec![]), // Optimize nothing
+            })
+            .await
+            .unwrap();
```
Comment (westonpace) [inline rust/lance/src/index.rs:1037]: Do you want to add a test case where `index_names` is set but non-empty?

**Review by westonpace — APPROVED** [review]
Looks fine but could maybe use one more test case

File: rust/lance/src/index.rs (hunk 2 above)
Comment (BubbleCal) [reply rust/lance/src/index.rs:1058]: good point, added tests to cover it

File: python/python/lance/dataset.py (hunk 1 above)
Comment (BubbleCal) [reply python/python/lance/dataset.py]: fixed
